File: src/main/java/syntax/java/DOMLegacy.java
package syntax.java;

import java.io.ByteArrayInputStream;
import java.io.InputStream;

import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;

public class DOMLegacy {
	public static void main(String[] args) throws Exception {
		String xhtml = "<html>"
				+"<h1> Oi </h1>"
				+"<h1> Mundo </h1>" 
				+"</html>";
		InputStream is = new ByteArrayInputStream(xhtml.getBytes());
		Document doc = DocumentBuilderFactory
	    		.newInstance()
	    		.newDocumentBuilder()
	    		.parse(is);
	    NodeList nodes = doc.getElementsByTagName("h1");
	    //Se ao menos NodeList estendesse Iterable...
	    for (int i = 0; i < nodes.getLength() ; i++) {
			String content = nodes.item(i).getTextContent();
			System.out.println(content);
		}
	}
}

File: src/main/scala/syntax/scala/DOMLegacy.scala
package syntax.scala

import java.io.ByteArrayInputStream
import javax.xml.parsers.DocumentBuilderFactory
import org.w3c.dom.NodeList
import org.w3c.dom.Node

import java.util

object DOMLegacy extends App {
  val xhtml = """<html>
    <h1> Oi </h1>
    <h1> Mundo </h1>
  </html>"""
  val is = new ByteArrayInputStream(xhtml.getBytes)
  val doc = DocumentBuilderFactory
              .newInstance
              .newDocumentBuilder
              .parse(is)
  val nodes:NodeList = doc.getElementsByTagName("h1")
	//Nada que uma conversao implicita nao resolva  
  implicit  class TraversableNodeList(nl: NodeList)
    extends Traversable[Node] {

    def iterator: Iterator[org.w3c.dom.Node] = ???

    override def foreach[U](f: Node => U): Unit =
      for (i <- 0 until nl.getLength) f(nl.item(i))
  }

  /*
  for (node <- nodes;
       content <- node.getTextContent)
    println(content)

   */
}
